### PR TITLE
Add unit tests for addon storyshots

### PR DESCRIPTION
**Pull request by tmeasday**
Issue:

We didn't have a good way to test storyshots.

## What I did

Added storyshots to `test-cra` app.

## How to test

```
npm install
npm run bootstrap
cd examples/test-cra
npm install
npm test
```

### CODE REVIEW

**Comment by tmeasday** [comment]
Note from commit:

> Storyshots wants to plugin to the version used by the app, not have its own version. So its a clear peer dep.

> Not sure it should depend on *storybook* at all given it is optional whether the user uses it or `react-native-storybook`

**Comment by ndelangen** [comment]
I pushed some fixed and updated your branch, hope you don't mind 👼

File: examples/cra-storybook/package.json [hunk 1]
```diff
@@ -4,7 +4,9 @@
   "private": true,
   "devDependencies": {
     "@kadira/storybook": "*",
-    "react-scripts": "0.9.5"
+    "react-scripts": "0.9.5",
+    "react-test-renderer": "^15.5.4",
+    "storyshots": "*"
```
Comment (ndelangen) [inline examples/cra-storybook/package.json]: Lerna will do it's thing on that!

**Comment by ndelangen** [comment]
Can this be used to help along https://github.com/storybooks/storybook/pull/896 ?

**Comment by tmeasday** [comment]
> Can this be used to help along #896 ?

I think #896 would be served by the alternate approach mentioned in https://github.com/storybooks/storybook/issues/917

**Comment by tmeasday** [comment]
Hmmm. This test is failing because I changed the dependency in storyshots to a peer dep, and the example project isn't using an addon.

I can see two solutions to the problem:

1. Return `@kadira/storybook-addons` to be a proper dep of storyshots (this still feels wrong as it is a plugin, and making it a direct dep is asking for versioning problems).

2. Follow this technique (https://github.com/storybooks/storybook/blob/9471a71933af45aa24c0db8d189c410809645b94/packages/storyshots/src/index.js#L15-L20) to check if the user has installed the addon package (i.e. is using any addons), and only mock if so.

I'm not sure of the best way to do 2. If you make the package a peer dep, and the user isn't using it, they will get a warning from npm, but things will work fine. This seems suboptimal though. The alternative is to just not specify anything in `package.json`.

Any advice @ndelangen?

**Comment by ndelangen** [comment]
I think not having it as a listed dependency is OK in this case as long as we guard when requiring it.

**Comment by ndelangen** [comment]
@tmeasday Do you think you can finish this now? Any more feedback needed? Let me know!

**Comment by tmeasday** [comment]
I'll get something going with that technique now.

**Comment by tmeasday** [comment]
> This test is failing because I changed the dependency in storyshots to a peer dep, and the example project isn't using an addon (so as a result addons never gets installed)

Actually this was wrong. Storybooks depends on addons by default. The issue is that due to nested `node_modules` structures in this development context, `storyshots` can't require the correct addons.

**Comment by tmeasday** [comment]
OK I think I finally resolved this by using the `test-cra` app. I suspect we should get rid of the confusing `cra-storybook` example @ndelangen.

**Comment by shilman** [comment]
Problems running `npm run bootstrap`:
```
lerna ERR! execute npm WARN enoent ENOENT: no such file or directory, open '/Users/shilman/projects/storybook/storybook/packages/addon-links/package.json'
lerna ERR! execute npm WARN react-storybook No repository field.
lerna ERR! execute npm WARN react-storybook No license field.
lerna ERR! execute /Users/shilman/projects/storybook/storybook/packages/react-storybook
```

File: packages/storyshots/src/index.js [hunk 2]
```diff
@@ -12,12 +12,16 @@ let configPath;
...
 
 const pkg = readPkgUp.sync().pkg;
-const isStorybook =
-  (pkg.devDependencies && pkg.devDependencies['@kadira/storybook']) ||
-  (pkg.dependencies && pkg.dependencies['@kadira/storybook']);
-const isRNStorybook =
-  (pkg.devDependencies && pkg.devDependencies['@kadira/react-native-storybook']) ||
-  (pkg.dependencies && pkg.dependencies['@kadira/react-native-storybook']);
+
+const hasDependency = function(name) {
```
Comment (shilman) [inline packages/storyshots/src/index.js]: 😻

**Review by shilman — APPROVED** [review]
~~Looks great!~~
~~Minor point: I prefer `renderOnly` to `justRender`, but it's all good.~~

OOPS WRONG PR.

**Comment by tmeasday** [comment]
> Minor point: I prefer renderOnly to justRender, but it's all good.

Isn't that the other PR? I think I prefer that too..

**Comment by tmeasday** [comment]
LMK if you want me to rebase this onto https://github.com/storybooks/storybook/pull/1031

**Comment by tmeasday** [comment]
@ndelangen Please!

**Comment by ndelangen** [comment]
pushed

**Comment by tmeasday** [comment]
Ok, I think I finally fixed this. The tests should pass 😬

**Comment by tmeasday** [comment]
See commit for more detail: https://github.com/storybooks/storybook/pull/971/commits/1c8f8e0ade576f1f5d5d19e4eb8ba680749a59e7

File: addons/storyshots/src/index.js [hunk 3]
```diff
@@ -12,16 +12,20 @@ let configPath;
...
+    (pkg.devDependencies && pkg.devDependencies[name]) ||
+    (pkg.dependencies && pkg.dependencies[name])
+  );
+};
 
 export default function testStorySnapshots(options = {}) {
   addons.setChannel(createChannel());
 
+  const isStorybook = options.mode === 'react' || hasDependency('@storybook/react');
+  const isRNStorybook = options.mode === 'react' || hasDependency('@storybook/react-native');
```
Comment (shilman) [inline addons/storyshots/src/index.js]: should be `react-native`?

**Review by shilman — CHANGES_REQUESTED** [review]
Does the `options.mode` flag need to be added to docs?

**Review by shilman — APPROVED** [review]

**Pull request merged**